File: code/06_geography_ibd_maps/run_ibd_excluding_MVZ216210_2026-03-02.py
#!/usr/bin/env python3
import os
import numpy as np
import pandas as pd
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
from scipy import stats

BASE = "/Users/gspellman/Trowbridgii_analyses/Stacks analysis of ddRAD data/stacks_refmap_sorex_excl_MVZ216210_2026-02-26_clean"
DOWN = os.path.join(BASE, "downstream_analysis_2026-02-23")
FIG = os.path.join(DOWN, "figures")
TAB = os.path.join(DOWN, "tables")
os.makedirs(FIG, exist_ok=True)
os.makedirs(TAB, exist_ok=True)

COORD = "/Users/gspellman/Trowbridgii_analyses/Sample_geographic_coordinants.txt"
POPMAP = os.path.join(BASE, "popmap.tsv")
VCF = os.path.join(BASE, "populations.snps.vcf")

OUT_STATS = os.path.join(TAB, "IBD_statistics_summary_excl_MVZ216210.tsv")
OUT_PAIRS = os.path.join(TAB, "IBD_pairwise_distances_excl_MVZ216210.tsv")
OUT_BINS = os.path.join(TAB, "IBD_distance_bin_summary_excl_MVZ216210.tsv")
OUT_SUMMARY = os.path.join(TAB, "IBD_analysis_summary_excl_MVZ216210.txt")

OUT_SCATTER_PNG = os.path.join(FIG, "IBD_scatter_regression_excl_MVZ216210_publication.png")
OUT_SCATTER_PDF = os.path.join(FIG, "IBD_scatter_regression_excl_MVZ216210_publication.pdf")
OUT_BIN_PNG = os.path.join(FIG, "IBD_distance_bin_trend_excl_MVZ216210_publication.png")
OUT_BIN_PDF = os.path.join(FIG, "IBD_distance_bin_trend_excl_MVZ216210_publication.pdf")


def parse_gt(gt_field):
    gt = gt_field.split(':', 1)[0].replace('|', '/')
    if gt in ("./.", "."):
        return np.nan
    if gt == "0/0":
        return 0.0
    if gt in ("0/1", "1/0"):
        return 1.0
    if gt == "1/1":
        return 2.0
    return np.nan


def read_vcf_matrix(vcf_path, sample_set):
    chrom, pos, vid, rows = [], [], [], []
    samples = None
    with open(vcf_path, 'r') as fh:
        sample_idx = None
        for ln in fh:
            if ln.startswith('##'):
                continue
            if ln.startswith('#CHROM'):
                h = ln.rstrip('\n').split('\t')
                all_samples = h[9:]
                samples = [s for s in all_samples if s in sample_set]
                sample_idx = [all_samples.index(s) for s in samples]
                continue
            p = ln.rstrip('\n').split('\t')
            ref, alt = p[3], p[4]
            if len(ref) != 1 or len(alt) != 1 or ',' in alt:
                continue
            g = p[9:]
            row = [parse_gt(g[j]) for j in sample_idx]
            chrom.append(p[0])
            pos.append(int(p[1]))
            vid.append(p[2] if p[2] != '.' else f"{p[0]}:{p[1]}")
            rows.append(row)
    return samples, np.array(chrom), np.array(pos), np.array(vid), np.array(rows, dtype=float)


def haversine_km(lat1, lon1, lat2, lon2):
    r = 6371.0088
    p1 = np.radians(lat1)
    p2 = np.radians(lat2)
    dphi = np.radians(lat2 - lat1)
    dl = np.radians(lon2 - lon1)
    a = np.sin(dphi / 2.0) ** 2 + np.cos(p1) * np.cos(p2) * np.sin(dl / 2.0) ** 2
    return 2.0 * r * np.arcsin(np.sqrt(a))


def mantel_test(d1, d2, perms=4999, seed=20260302):
    rng = np.random.default_rng(seed)
    n = d1.shape[0]
    triu = np.triu_indices(n, k=1)
    x = d1[triu]
    y = d2[triu]
    r_obs = np.corrcoef(x, y)[0, 1]
    c = 0
    for _ in range(perms):
        p = rng.permutation(n)
        yp = d2[p][:, p][triu]
        r = np.corrcoef(x, yp)[0, 1]
        if abs(r) >= abs(r_obs):
            c += 1
    pval = (c + 1) / (perms + 1)
    return r_obs, pval


def main():
    for req in (COORD, POPMAP, VCF):
        if not os.path.exists(req):
            raise FileNotFoundError(req)

    coords = pd.read_csv(COORD, sep='\t')
    coords = coords.dropna(subset=['Sample', 'Latitude', 'Longitude']).copy()
    coords['Sample'] = coords['Sample'].astype(str)

    pop = pd.read_csv(POPMAP, sep='\t', header=None, names=['Sample', 'Population'])
    pop['Sample'] = pop['Sample'].astype(str)

    if 'MVZ216210' in set(pop['Sample']):
        raise RuntimeError('MVZ216210 still present in popmap for excluded dataset.')

    meta = pop.merge(coords, on='Sample', how='inner').copy()
    meta = meta.sort_values(['Population', 'Sample']).reset_index(drop=True)

    sample_set = set(meta['Sample'])
    samples, _chrom, _pos, _vid, G = read_vcf_matrix(VCF, sample_set)
    if samples is None or len(samples) < 3:
        raise RuntimeError('Insufficient samples parsed from VCF for IBD analysis.')

    # align metadata to genotype sample order
    meta2 = meta.set_index('Sample').loc[samples].reset_index()

    # standard SNP filters for distance analysis
    site_missing = np.mean(np.isnan(G), axis=1)
    p = np.nanmean(G, axis=1) / 2.0
    maf = np.minimum(p, 1.0 - p)
    keep = (site_missing <= 0.20) & np.isfinite(maf) & (maf >= 0.05)
    G = G[keep]
    if G.shape[0] < 10:
        raise RuntimeError('Too few SNPs remain after filtering for IBD analysis.')

    n = len(samples)
    geo = np.zeros((n, n), dtype=float)
    gen = np.zeros((n, n), dtype=float)
    pair_rows = []

    for i in range(n - 1):
        gi = G[:, i]
        for j in range(i + 1, n):
            gj = G[:, j]
            ok = np.isfinite(gi) & np.isfinite(gj)
            if np.any(ok):
                dgen = float(np.mean(np.abs(gi[ok] - gj[ok]) / 2.0))
            else:
                dgen = np.nan

            dgeo = float(haversine_km(
                meta2.loc[i, 'Latitude'], meta2.loc[i, 'Longitude'],
                meta2.loc[j, 'Latitude'], meta2.loc[j, 'Longitude']
            ))

            gen[i, j] = gen[j, i] = dgen
            geo[i, j] = geo[j, i] = dgeo

            pair_rows.append({
                'Sample1': samples[i],
                'Sample2': samples[j],
                'Population1': meta2.loc[i, 'Population'],
                'Population2': meta2.loc[j, 'Population'],
                'Geographic_distance_km': dgeo,
                'Genetic_distance': dgen,
            })

    mgen = np.nanmedian(gen[np.isfinite(gen)])
    gen[~np.isfinite(gen)] = mgen

    triu = np.triu_indices(n, k=1)
    x = geo[triu]
    y = gen[triu]

    pear_r, pear_p = stats.pearsonr(x, y)
    spear_r, spear_p = stats.spearmanr(x, y)
    mantel_r, mantel_p = mantel_test(gen, geo, perms=4999)

    lm = stats.linregress(x, y)

    stats_df = pd.DataFrame([{
        'Dataset': 'stacks_refmap_sorex_excl_MVZ216210',
        'Excluded_sample': 'MVZ216210',
        'N_samples': n,
        'N_pairwise_comparisons': len(x),
        'N_SNPs_after_filtering': int(G.shape[0]),
        'Pearson_r': float(pear_r),
        'Pearson_p': float(pear_p),
        'Spearman_rho': float(spear_r),
        'Spearman_p': float(spear_p),
        'Mantel_r': float(mantel_r),
        'Mantel_p_permutation': float(mantel_p),
        'Mantel_permutations': 4999,
        'Linear_slope': float(lm.slope),
        'Linear_intercept': float(lm.intercept),
        'Linear_r2': float(lm.rvalue ** 2),
        'Linear_p': float(lm.pvalue),
        'Linear_slope_stderr': float(lm.stderr),
    }])
    stats_df.to_csv(OUT_STATS, sep='\t', index=False)

    pairs_df = pd.DataFrame(pair_rows)
    pairs_df.to_csv(OUT_PAIRS, sep='\t', index=False)

    bins = np.quantile(x, np.linspace(0, 1, 11))
    bins = np.unique(bins)
    idx = np.digitize(x, bins[1:-1], right=True)
    b_rows = []
    for b in range(len(bins) - 1):
        m = idx == b
        if np.sum(m) == 0:
            continue
        b_rows.append({
            'Bin': b + 1,
            'Geo_km_min': float(bins[b]),
            'Geo_km_max': float(bins[b + 1]),
            'N_pairs': int(np.sum(m)),
            'Mean_genetic_distance': float(np.mean(y[m])),
            'SE_genetic_distance': float(np.std(y[m], ddof=1) / np.sqrt(np.sum(m))) if np.sum(m) > 1 else np.nan,
        })
    bin_df = pd.DataFrame(b_rows)
    bin_df.to_csv(OUT_BINS, sep='\t', index=False)

    for out in (OUT_SCATTER_PNG, OUT_SCATTER_PDF):
        plt.figure(figsize=(8.2, 6.4), dpi=400 if out.endswith('.png') else None)
        ax = plt.gca()
        ax.scatter(x, y, s=16, c='#4C78A8', alpha=0.40, edgecolors='none')
        xs = np.linspace(np.nanmin(x), np.nanmax(x), 300)
        ax.plot(xs, lm.intercept + lm.slope * xs, color='#D62728', linewidth=2.1)
        txt = (
            f"Pearson r = {pear_r:.3f} (p = {pear_p:.2e})\\n"
            f"Spearman rho = {spear_r:.3f} (p = {spear_p:.2e})\\n"
            f"Mantel r = {mantel_r:.3f} (p = {mantel_p:.4f}, 4,999 perms)\\n"
            f"Linear slope = {lm.slope:.3e}"
        )
        ax.text(
            0.02, 0.98, txt, transform=ax.transAxes, va='top', ha='left', fontsize=10,
            bbox=dict(facecolor='white', alpha=0.88, edgecolor='#BFBFBF')
        )
        ax.set_xlabel('Geographic distance (km)', fontsize=11)
        ax.set_ylabel('Genetic distance (mean allele-dosage difference / 2)', fontsize=11)
        ax.set_title('Isolation by distance (MVZ216210 excluded)', fontsize=12)
        ax.grid(color='#E0E0E0', linestyle=':', linewidth=0.9)
        plt.tight_layout()
        plt.savefig(out)
        plt.close()

    for out in (OUT_BIN_PNG, OUT_BIN_PDF):
        plt.figure(figsize=(7.8, 5.6), dpi=400 if out.endswith('.png') else None)
        xc = (bin_df['Geo_km_min'] + bin_df['Geo_km_max']) / 2.0
        plt.errorbar(
            xc, bin_df['Mean_genetic_distance'], yerr=bin_df['SE_genetic_distance'],
            fmt='o-', color='#1F77B4', ecolor='#1F77B4', elinewidth=1.4, capsize=3.0, linewidth=1.6
        )
        plt.xlabel('Geographic distance bin midpoint (km)', fontsize=11)
        plt.ylabel('Mean genetic distance', fontsize=11)
        plt.title('IBD distance-class trend (MVZ216210 excluded)', fontsize=12)
        plt.grid(color='#E0E0E0', linestyle=':', linewidth=0.9)
        plt.tight_layout()
        plt.savefig(out)
        plt.close()

    with open(OUT_SUMMARY, 'w') as fh:
        fh.write('Isolation-by-distance analysis summary (MVZ216210 excluded)\n')
        fh.write(f'Samples analyzed: {n}\n')
        fh.write(f'SNPs after filters: {G.shape[0]}\n')
        fh.write(f'Pairwise comparisons: {len(x)}\n')
        fh.write(f'Pearson r={pear_r:.6f}, p={pear_p:.6e}\n')
        fh.write(f'Spearman rho={spear_r:.6f}, p={spear_p:.6e}\n')
        fh.write(f'Mantel r={mantel_r:.6f}, p={mantel_p:.6f}, permutations=4999\n')
        fh.write(f'Scatter figure: {OUT_SCATTER_PNG}\n')
        fh.write(f'Binned figure: {OUT_BIN_PNG}\n')
        fh.write(f'Stats table: {OUT_STATS}\n')
        fh.write(f'Pairwise table: {OUT_PAIRS}\n')
        fh.write(f'Binned table: {OUT_BINS}\n')

    print('IBD analysis complete', flush=True)


if __name__ == '__main__':
    main()

File: code/06_geography_ibd_maps/run_ibd_excluding_MVZ216210_2026-03-02.R
#!/usr/bin/env Rscript
options(stringsAsFactors = FALSE)

base_dir <- "/Users/gspellman/Trowbridgii_analyses/Stacks analysis of ddRAD data/stacks_refmap_sorex_excl_MVZ216210_2026-02-26_clean"
down_dir <- file.path(base_dir, "downstream_analysis_2026-02-23")
fig_dir <- file.path(down_dir, "figures")
tab_dir <- file.path(down_dir, "tables")
dir.create(fig_dir, recursive = TRUE, showWarnings = FALSE)
dir.create(tab_dir, recursive = TRUE, showWarnings = FALSE)

coord_file <- "/Users/gspellman/Trowbridgii_analyses/Sample_geographic_coordinants.txt"
popmap_file <- file.path(base_dir, "popmap.tsv")
vcf_file <- file.path(base_dir, "populations.snps.vcf")

out_stats <- file.path(tab_dir, "IBD_statistics_summary_excl_MVZ216210.tsv")
out_pairs <- file.path(tab_dir, "IBD_pairwise_distances_excl_MVZ216210.tsv")
out_bins <- file.path(tab_dir, "IBD_distance_bin_summary_excl_MVZ216210.tsv")
out_summary <- file.path(tab_dir, "IBD_analysis_summary_excl_MVZ216210.txt")
out_png <- file.path(fig_dir, "IBD_scatter_regression_excl_MVZ216210_publication.png")
out_pdf <- file.path(fig_dir, "IBD_scatter_regression_excl_MVZ216210_publication.pdf")
out_bin_png <- file.path(fig_dir, "IBD_distance_bin_trend_excl_MVZ216210_publication.png")
out_bin_pdf <- file.path(fig_dir, "IBD_distance_bin_trend_excl_MVZ216210_publication.pdf")

logmsg <- function(...) cat(sprintf("[%s] ", format(Sys.time(), "%Y-%m-%d %H:%M:%S")), ..., "\n", sep = "")
stop_if_missing <- function(x) {
  m <- x[!file.exists(x)]
  if (length(m) > 0) stop("Missing required files:\n", paste(m, collapse = "\n"))
}

parse_gt <- function(x) {
  gt <- sub(":.*$", "", x)
  gt <- gsub("\\|", "/", gt)
  out <- rep(NA_real_, length(gt))
  out[gt == "0/0"] <- 0
  out[gt %in% c("0/1", "1/0")] <- 1
  out[gt == "1/1"] <- 2
  out
}

haversine_km <- function(lat1, lon1, lat2, lon2) {
  r <- 6371.0088
  p1 <- lat1 * pi / 180
  p2 <- lat2 * pi / 180
  dphi <- (lat2 - lat1) * pi / 180
  dl <- (lon2 - lon1) * pi / 180
  a <- sin(dphi / 2)^2 + cos(p1) * cos(p2) * sin(dl / 2)^2
  2 * r * asin(sqrt(a))
}

mantel_test <- function(d1, d2, perms = 4999, seed = 20260302) {
  set.seed(seed)
  n <- nrow(d1)
  idx <- upper.tri(d1)
  x <- d1[idx]
  y <- d2[idx]
  r_obs <- suppressWarnings(cor(x, y, method = "pearson"))
  more <- 0L
  for (i in seq_len(perms)) {
    p <- sample.int(n)
    yp <- d2[p, p][idx]
    r <- suppressWarnings(cor(x, yp, method = "pearson"))
    if (is.finite(r) && abs(r) >= abs(r_obs)) more <- more + 1L
  }
  pval <- (more + 1) / (perms + 1)
  c(r = r_obs, p = pval)
}

stop_if_missing(c(coord_file, popmap_file, vcf_file))

logmsg("Loading coordinates and popmap")
coord <- read.delim(coord_file, header = TRUE, sep = "\t", check.names = FALSE)
pop <- read.delim(popmap_file, header = FALSE, sep = "\t", col.names = c("Sample", "Population"), check.names = FALSE)
coord$Sample <- as.character(coord$Sample)
pop$Sample <- as.character(pop$Sample)

if ("MVZ216210" %in% pop$Sample) stop("MVZ216210 found in excluded dataset popmap")

meta <- merge(pop, coord[, c("Sample", "Latitude", "Longitude")], by = "Sample", all = FALSE)
meta <- meta[order(meta$Population, meta$Sample), ]
meta$Latitude <- as.numeric(as.character(meta$Latitude))
meta$Longitude <- as.numeric(as.character(meta$Longitude))
if (any(!is.finite(meta$Latitude)) || any(!is.finite(meta$Longitude))) {
  stop("Non-finite coordinates detected after numeric conversion.")
}
if (min(meta$Latitude) < -90 || max(meta$Latitude) > 90 || min(meta$Longitude) < -180 || max(meta$Longitude) > 180) {
  stop("Coordinate ranges are invalid; check latitude/longitude parsing.")
}

sample_set <- unique(meta$Sample)

logmsg("Parsing VCF and extracting genotypes for matched samples")
lns <- readLines(vcf_file, warn = FALSE)
header_line <- lns[grep("^#CHROM", lns)]
if (length(header_line) != 1) stop("VCF #CHROM header not found")
h <- strsplit(header_line, "\t", fixed = TRUE)[[1]]
all_samples <- h[10:length(h)]
keep_samples <- all_samples[all_samples %in% sample_set]
if (length(keep_samples) < 3) stop("Too few matched samples in VCF")
keep_idx <- match(keep_samples, all_samples)

variant_lines <- lns[!grepl("^#", lns)]
G_list <- vector("list", length(variant_lines))
keep_row <- logical(length(variant_lines))
ri <- 0L
for (k in seq_along(variant_lines)) {
  p <- strsplit(variant_lines[k], "\t", fixed = TRUE)[[1]]
  ref <- p[4]
  alt <- p[5]
  if (nchar(ref) != 1 || nchar(alt) != 1 || grepl(",", alt, fixed = TRUE)) next
  gfields <- p[10:length(p)]
  gt <- parse_gt(gfields[keep_idx])
  ri <- ri + 1L
  G_list[[ri]] <- gt
  keep_row[ri] <- TRUE
}
G_list <- G_list[seq_len(ri)]
G <- do.call(rbind, G_list)
colnames(G) <- keep_samples

# Align meta order to genotype order
meta2 <- meta[match(keep_samples, meta$Sample), ]
if (any(!is.finite(meta2$Latitude)) || any(!is.finite(meta2$Longitude))) {
  stop("Failed to align finite coordinates to genotype sample order.")
}

# Preflight geographic sanity check to prevent coordinate-parsing artifacts.
geo_check <- matrix(0, nrow(meta2), nrow(meta2))
for (ii in seq_len(nrow(meta2) - 1)) {
  for (jj in (ii + 1):nrow(meta2)) {
    dchk <- haversine_km(meta2$Latitude[ii], meta2$Longitude[ii], meta2$Latitude[jj], meta2$Longitude[jj])
    geo_check[ii, jj] <- dchk
    geo_check[jj, ii] <- dchk
  }
}
max_geo_check <- max(geo_check, na.rm = TRUE)
if (max_geo_check > 2000) {
  stop(sprintf("Geographic distance sanity check failed (max=%.3f km > 2000 km).", max_geo_check))
}

logmsg("Applying SNP filters (missing<=20%, MAF>=0.05)")
site_missing <- rowMeans(is.na(G))
p <- rowMeans(G, na.rm = TRUE) / 2
maf <- pmin(p, 1 - p)
keep <- is.finite(site_missing) & is.finite(maf) & site_missing <= 0.20 & maf >= 0.05
Gf <- G[keep, , drop = FALSE]
if (nrow(Gf) < 10) stop("Too few SNPs remain after filtering")

logmsg("Computing pairwise geographic and genetic distances")
n <- ncol(Gf)
geo <- matrix(0, n, n)
gen <- matrix(0, n, n)
pairs <- vector("list", n * (n - 1) / 2)
pair_i <- 0L
for (i in seq_len(n - 1)) {
  gi <- Gf[, i]
  for (j in (i + 1):n) {
    gj <- Gf[, j]
    ok <- is.finite(gi) & is.finite(gj)
    dgen <- if (any(ok)) mean(abs(gi[ok] - gj[ok]) / 2) else NA_real_
    dgeo <- haversine_km(meta2$Latitude[i], meta2$Longitude[i], meta2$Latitude[j], meta2$Longitude[j])
    gen[i, j] <- dgen
    gen[j, i] <- dgen
    geo[i, j] <- dgeo
    geo[j, i] <- dgeo

    pair_i <- pair_i + 1L
    pairs[[pair_i]] <- data.frame(
      Sample1 = keep_samples[i],
      Sample2 = keep_samples[j],
      Population1 = meta2$Population[i],
      Population2 = meta2$Population[j],
      Geographic_distance_km = dgeo,
      Genetic_distance = dgen,
      stringsAsFactors = FALSE
    )
  }
}

mgen <- median(gen[is.finite(gen)], na.rm = TRUE)
gen[!is.finite(gen)] <- mgen

idx <- upper.tri(gen)
x <- geo[idx]
y <- gen[idx]

logmsg("Running significance tests")
pear <- cor.test(x, y, method = "pearson")
spea <- cor.test(x, y, method = "spearman", exact = FALSE)
man <- mantel_test(gen, geo, perms = 4999, seed = 20260302)
lmfit <- lm(y ~ x)
lms <- summary(lmfit)

stats_df <- data.frame(
  Dataset = "stacks_refmap_sorex_excl_MVZ216210",
  Excluded_sample = "MVZ216210",
  N_samples = n,
  N_pairwise_comparisons = length(x),
  N_SNPs_after_filtering = nrow(Gf),
  Pearson_r = unname(pear$estimate),
  Pearson_p = pear$p.value,
  Spearman_rho = unname(spea$estimate),
  Spearman_p = spea$p.value,
  Mantel_r = unname(man["r"]),
  Mantel_p_permutation = unname(man["p"]),
  Mantel_permutations = 4999,
  Linear_slope = coef(lmfit)[2],
  Linear_intercept = coef(lmfit)[1],
  Linear_r2 = lms$r.squared,
  Linear_p = coef(lms)[2, 4],
  Linear_slope_stderr = coef(lms)[2, 2],
  check.names = FALSE
)
write.table(stats_df, out_stats, sep = "\t", row.names = FALSE, quote = FALSE)

pairs_df <- do.call(rbind, pairs)
write.table(pairs_df, out_pairs, sep = "\t", row.names = FALSE, quote = FALSE)

bins <- as.numeric(quantile(x, probs = seq(0, 1, 0.1), na.rm = TRUE))
bins <- unique(bins)
bin_id <- cut(x, breaks = bins, include.lowest = TRUE, labels = FALSE)
bin_rows <- list()
for (b in seq_len(length(bins) - 1)) {
  m <- which(bin_id == b)
  if (length(m) == 0) next
  yy <- y[m]
  bin_rows[[length(bin_rows) + 1L]] <- data.frame(
    Bin = b,
    Geo_km_min = bins[b],
    Geo_km_max = bins[b + 1],
    N_pairs = length(m),
    Mean_genetic_distance = mean(yy, na.rm = TRUE),
    SE_genetic_distance = if (length(yy) > 1) sd(yy, na.rm = TRUE) / sqrt(length(yy)) else NA_real_,
    check.names = FALSE
  )
}
bin_df <- do.call(rbind, bin_rows)
write.table(bin_df, out_bins, sep = "\t", row.names = FALSE, quote = FALSE)

logmsg("Rendering publication IBD figures")
plot_scatter <- function(outf, is_png = TRUE) {
  if (is_png) png(outf, width = 3200, height = 2500, res = 400) else pdf(outf, width = 8.0, height = 6.25, useDingbats = FALSE)
  par(mar = c(5.0, 5.2, 3.2, 1.2))
  plot(x, y, pch = 16, col = rgb(76/255, 120/255, 168/255, 0.40), cex = 0.75,
       xlab = "Geographic distance (km)",
       ylab = "Genetic distance (mean allele-dosage difference / 2)",
       main = "Isolation by distance (MVZ216210 excluded)")
  abline(lmfit, col = "#D62728", lwd = 2.3)
  grid(col = "#E0E0E0", lty = 3)
  legend("topleft", bty = "n", cex = 0.90, text.col = "#202020",
         legend = c(
           sprintf("Pearson r = %.3f (p = %.2e)", unname(pear$estimate), pear$p.value),
           sprintf("Spearman rho = %.3f (p = %.2e)", unname(spea$estimate), spea$p.value),
           sprintf("Mantel r = %.3f (p = %.4f, 4,999 perms)", unname(man["r"]), unname(man["p"]))
         ))
  dev.off()
}

plot_scatter(out_png, TRUE)
plot_scatter(out_pdf, FALSE)

plot_bins <- function(outf, is_png = TRUE) {
  if (is_png) png(outf, width = 3000, height = 2200, res = 400) else pdf(outf, width = 7.5, height = 5.5, useDingbats = FALSE)
  par(mar = c(5.0, 5.2, 3.2, 1.2))
  xc <- (bin_df$Geo_km_min + bin_df$Geo_km_max) / 2
  plot(xc, bin_df$Mean_genetic_distance, type = "b", pch = 16, cex = 0.9, lwd = 1.8,
       col = "#1F77B4",
       xlab = "Geographic distance bin midpoint (km)",
       ylab = "Mean genetic distance",
       main = "IBD distance-class trend (MVZ216210 excluded)")
  arrows(x0 = xc, y0 = bin_df$Mean_genetic_distance - bin_df$SE_genetic_distance,
         x1 = xc, y1 = bin_df$Mean_genetic_distance + bin_df$SE_genetic_distance,
         angle = 90, code = 3, length = 0.03, col = "#1F77B4", lwd = 1.3)
  grid(col = "#E0E0E0", lty = 3)
  dev.off()
}

plot_bins(out_bin_png, TRUE)
plot_bins(out_bin_pdf, FALSE)

writeLines(c(
  "Isolation-by-distance analysis summary (MVZ216210 excluded)",
  paste0("Samples analyzed: ", n),
  paste0("SNPs after filters: ", nrow(Gf)),
  paste0("Pairwise comparisons: ", length(x)),
  sprintf("Max pairwise geographic distance (sanity check): %.3f km", max_geo_check),
  sprintf("Pearson r = %.6f, p = %.6e", unname(pear$estimate), pear$p.value),
  sprintf("Spearman rho = %.6f, p = %.6e", unname(spea$estimate), spea$p.value),
  sprintf("Mantel r = %.6f, p = %.6f, permutations = 4999", unname(man["r"]), unname(man["p"])),
  paste0("Scatter figure: ", out_png),
  paste0("Distance-bin figure: ", out_bin_png),
  paste0("Stats table: ", out_stats),
  paste0("Pairwise table: ", out_pairs),
  paste0("Bin table: ", out_bins)
), con = out_summary)

logmsg("IBD analysis complete")
